### PR TITLE
fix: SamFunctionProvider could miss functions with same name but in different stacks

### PR DESCRIPTION
**Pull request by aahung**
#### Which issue(s) does this change fix?

This is a follow up of #2594, 2nd PR to support nested stack.
*These two commits are taken out from https://github.com/aahung/aws-sam-cli/tree/nested-stack-2

#### Why is this change necessary?

two functions with the same name can appear in two different nested stacks, which causing one overriding the other. 

#### How does it address the issue?

Before the dict's key is function name, which is not unique in a multi-stack setting. Now the key is a unique key "full_path" consisting of stack_path and function name.

#### What side effects does this change have?

N/A, nested stack support is not available and this PR has no customer impact.

#### Checklist

- [ ] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [x] Write unit tests
- [ ] Write/update functional tests
- [ ] Write/update integration tests
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

File: samcli/lib/providers/sam_function_provider.py [hunk 1]
```diff
@@ -108,7 +108,7 @@ def _extract_functions(stacks: List[Stack], ignore_code_extraction_warnings=Fals
             Function configuration object
         """
 
-        result = {}
+        result = {}  # dict: full_path -> Function
```
Comment (c2tarun) [inline samcli/lib/providers/sam_function_provider.py:111]: no types?

File: samcli/lib/providers/sam_function_provider.py [hunk 2]
```diff
@@ -126,13 +126,14 @@ def _extract_functions(stacks: List[Stack], ignore_code_extraction_warnings=Fals
                         stack.resources,
                         ignore_code_extraction_warnings=ignore_code_extraction_warnings,
                     )
-                    result[name] = SamFunctionProvider._convert_sam_function_resource(
+                    f = SamFunctionProvider._convert_sam_function_resource(
```
Comment (c2tarun) [inline samcli/lib/providers/sam_function_provider.py:129]: f -> function

**Review by c2tarun — CHANGES_REQUESTED** [review]

File: samcli/lib/providers/sam_function_provider.py (hunk 1 above)
Comment (aahung) [reply samcli/lib/providers/sam_function_provider.py]: Sorry about the confusion, I meant to explain what are the keys and values. Updated

File: samcli/lib/providers/sam_function_provider.py (hunk 2 above)
Comment (aahung) [reply samcli/lib/providers/sam_function_provider.py]: renamed.

**Review by mndeveci — APPROVED** [review]
LGTM. Quick question, if I want to invoke a function from child stack, can I invoke it by running `sam local invoke ChildStack/Function`? Same question applies for `sam build` as well.

**Comment by aahung** [comment]
> LGTM. Quick question, if I want to invoke a function from child stack, can I invoke it by running `sam local invoke ChildStack/Function`? Same question applies for `sam build` as well.

Yes, but not yet supported until implemented in next PR, we can either invoke by `local invoke Function` or `local invoke ChildStack/Function` (to avoid ambiguity).

**Review by c2tarun — APPROVED** [review]

File: samcli/lib/providers/provider.py [hunk 3]
```diff
@@ -371,3 +393,11 @@ def resources(self) -> Dict:
...
         resources: Dict = processed_template_dict.get("Resources", {})
         return resources
+
+
+def get_full_path(stack_path: str, logical_id: str) -> str:
+    """
+    Return the unique posix path-like identifier
+    while will used for identify a resource from resources in a multi-stack situation
+    """
+    return posixpath.join(stack_path, logical_id)
```
Comment (sriram-mv) [inline samcli/lib/providers/provider.py:403]: are there any concerns here, where posixpath might behave weirdly trying to look for absolute paths and discard anything else?

**Review by sriram-mv — CHANGES_REQUESTED** [review]

File: samcli/lib/providers/provider.py (hunk 3 above)
Comment (aahung) [reply samcli/lib/providers/provider.py:403]: this `full_path` will not be used in the file system, it is a `posix path-like identifier` used in `sam build` and also in `sam local invoke <full_path>`

Comment (sriram-mv) [reply samcli/lib/providers/provider.py:403]: ok sounds good, the reason I ask is sometimes Pathlib on windows tries to resolve paths before joining them, I wanted to make sure that posixpath does not run into the same issue.

Comment (aahung) [reply samcli/lib/providers/provider.py:403]: Here is the source code of `posixpath.join`, we are pretty safe

https://github.com/python/cpython/blob/fd053fdd39fbdf114b4218ea4309666bafa95788/Lib/posixpath.py#L67-L92

**Review by sriram-mv — APPROVED** [review]

**Pull request merged**